In [62]:
!pip install sqlalchemy
!pip install psycopg2

In [2]:
from sqlalchemy import create_engine

## Question 1

In [61]:
# Database credentials
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'dvdrentals'

# use the credentials to start a connection
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

sql = '''
select rating, count(*)
from film
group by rating '''
film_count_by_rating = engine.execute(sql)
rows = film_count_by_rating.fetchall()
print(rows)

[('R', 195), ('NC-17', 210), ('PG', 194), ('PG-13', 223), ('G', 178)]


### Answer (question 1)
- 195 rated R films
- 210 NC-17 films
- 223 PG-13 films
- 194 PG films
- 178 rate G films

## Question 2

In [60]:
sql = '''
select rating, avg(rental_duration)
from film
group by rating '''
avg_rental_duration = engine.execute(sql)
rows = avg_rental_duration.fetchall()
print(rows)

[('R', Decimal('4.7743589743589744')), ('NC-17', Decimal('5.1428571428571429')), ('PG', Decimal('5.0824742268041237')), ('PG-13', Decimal('5.0538116591928251')), ('G', Decimal('4.8370786516853933'))]


### Answer (question 2)
Rated R films - 4.77 average rental duration <br>
Rated PG-13 films - 5.05 average rental duration <br>
Rated PG films - 5.08 average rental duration <br>
Rated G films - 4.84 average rental duration <br>
Rated NC-17 films - 5.14 average rental duration

## Question 3

In [53]:
sql = '''
select length
from film '''
film_lengths = engine.execute(sql)
rows = film_lengths.fetchall()
film_lengths = [x['length'] for x in rows]
film_lengths

def mean(nums):
    total = 0
    for num in nums:
        total += num
    count = len(nums)
    avg = total/count
    return avg
mean(film_lengths)

115.272

### Answer (question 3)
mean film length = 115.27

## Question 4

In [47]:
def median(nums):
    nums.sort()
    if len(nums)%2 == 0:
        return (nums[int(len(nums)/2 - 1)] + nums[int(len(nums)/2)])/2
    else:
        return nums[int(len(nums)/2)]

### Answer (question 4)
Median film length = 114

## Question 5

In [55]:
def std_dev(nums):
    avg = mean(nums)
    total = 0
    for num in nums:
        diff = (num - avg)**2
        total += diff
    return (total/(len(nums) - 1))**.5
std_dev(film_lengths)      

40.42633181855988

### Answer (question 5)
Standard deviation = 40.43